wgengine/router/router_linux.go

```diff
@@ -478,7 +478,42 @@ func (r *linuxRouter) delRoute(cidr netaddr.IPPrefix) error {
 	if r.ipRuleAvailable {
 		args = append(args, "table", tailscaleRouteTable)
 	}
-	return r.cmd.run(args...)
+	err := r.cmd.run(args...)
+	if err != nil {
+		ok, err := r.hasRoute(cidr)
+		if err != nil {
+			r.logf("warning: error checking whether %v even exists after error deleting it: %v", err)
+		} else {
+			if !ok {
+				r.logf("warning: tried to delete route %v but it was already gone; ignoring error", cidr)
+				return nil
+			}
+		}
+	}
+	return err
+}
+
+func dashFam(ip netaddr.IP) string {
+	if ip.Is6() {
+		return "-6"
+	}
+	return "-4"
+}
+
+func (r *linuxRouter) hasRoute(cidr netaddr.IPPrefix) (bool, error) {
+	args := []string{
+		"ip", dashFam(cidr.IP), "route", "show",
+		normalizeCIDR(cidr),
+		"dev", r.tunname,
+	}
+	if r.ipRuleAvailable {
+		args = append(args, "table", tailscaleRouteTable)
+	}
+	out, err := r.cmd.output(args...)
+	if err != nil {
+		return false, err
+	}
+	return len(out) > 0, nil
```

Comment:
It does:

```
root@tsdev:/home/bradfitz/src/tailscale.com# docker run --privileged -ti busybox ip route foo ; echo $?
ip: invalid argument 'foo' to 'ip'
1
root@tsdev:/home/bradfitz/src/tailscale.com# docker run --privileged -ti busybox ip route  ; echo $?
default via 172.17.0.1 dev eth0 
172.17.0.0/16 dev eth0 scope link  src 172.17.0.2 
0
root@tsdev:/home/bradfitz/src/tailscale.com# docker run --privileged -ti busybox ip route show 172.17.0.0/16  ; echo $?
172.17.0.0/16 dev eth0 scope link  src 172.17.0.2 
0
root@tsdev:/home/bradfitz/src/tailscale.com# docker run --privileged -ti busybox ip route show 172.17.0.2/32  ; echo $?
0
root@tsdev:/home/bradfitz/src/tailscale.com# 
```

Reply:
Might be worth hooking up to netlink instead of going through `ip` to avoid those types of headaches. I don't think it's much worse, and then we can actually maintain an accurate cache since we'll get updates when routes change.

Reply:
Indeed. That's what wgengine/monitor does. We already parse most of those message these days.